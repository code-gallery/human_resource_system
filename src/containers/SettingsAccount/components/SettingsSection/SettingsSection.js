import React from 'react'
import PropTypes from 'prop-types'
import Select from 'react-select'
import Form from 'components/Abstract/Form'
import ProfileVisibility from '../ProfileVisibility'

class SettingsSection extends Form {
  constructor(props) {
    super(props)
    this.renderSection = this.renderSection.bind(this)
    this.state = {
      ...props.settings,
      profileVisibility: (props.user) ? props.user.visibility : void 0
    }
  }

  save = (e) => {
    e.preventDefault()
    const { type, user, updateUserProfile } = this.props
    const data = { ...this.state }
    const profileVisibility = data.profileVisibility
    delete data.profileVisibility

    this.props.saveUserSettings(data)

    if (type === 'privacy' && user && profileVisibility && profileVisibility !== user.visibility) {
      updateUserProfile({ visibility: profileVisibility })
    }
  }

  renderInput(item, key) {
    if (item[key].type === 'enum') {
      const value = this.state[key] || 'registered'
      const options = []
      for (var prop in item[key].enum) {
        options.push({ value: prop, label: item[key].enum[prop] })
      }

      return (
        <div className="pull-right col-xs-12 col-md-3">
          <Select
            name={`select-${key}`}
            placeholder="Select setting"
            options={options}
            value={value}
            onChange={(obj) => {
              this.onChangeInput({
                target: {
                  name: key,
                  value: obj.value
                }
              })
            }}
          />
        </div>
      )
    } else {
      return (
        <div className="form-checkbox no-label pull-right">
          <input
            type="checkbox"
            id={key}
            name={key}
            checked={this.state[key]}
            onChange={this.onChangeInput}
          />
          <label htmlFor={key}>&nbsp;</label>
        </div>
      )
    }
  }

  renderSection(item, key) {
    const { data } = this.props

    return (
      <div className="form-group" key={key}>
        <h2 className="form-group">{data.sections[item].title}</h2>
        <ul>
          {
            Object.keys(data.sections[item].settings).map((key, index) => {
              const label = data.sections[item].settings[key].description

              return (
                <li key={index}>
                  <div className="form-group text-center">
                    <span className="pull-left settings-item-text">{label}</span>
                    {this.renderInput(data.sections[item].settings, key)}
                  </div>
                </li>
              )
            })
          }
        </ul>
      </div>
    )
  }

  render() {
    const { data, type } = this.props
    const sections = Object.keys(data.sections)
    const showProfileVisibility = (type === 'privacy')

    if (sections.length === 0) {
      return null
    }

    return (
      <div>
        <form>
          {sections.map(this.renderSection)}
          {showProfileVisibility &&
            <ProfileVisibility
              value={this.state.profileVisibility}
              onChangeInput={this.onChangeInput}
            />
          }
        </form>
        <button className="form-btn save" onClick={this.save}>
          Save
        </button>
      </div>
    )
  }
}

SettingsSection.propTypes = {
  type: PropTypes.string.isRequired,
  user: PropTypes.object,
  data: PropTypes.object.isRequired,
  settings: PropTypes.object.isRequired,
  saveUserSettings: PropTypes.func.isRequired,
  updateUserProfile: PropTypes.func.isRequired
}

export default SettingsSection
