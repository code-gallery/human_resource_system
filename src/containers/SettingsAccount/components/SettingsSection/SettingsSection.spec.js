import React from 'react'
import renderer from 'react-test-renderer'
import SettingsSection from './SettingsSection'

describe('containers/SettingsAccount/components/SettingsSection', () => {
  describe('when type = communications', () => {
    const settingData = {
      sections: {
        email_frequency: {
          title: 'Email Frequency',
          settings: {
            email_invite_verify: {
              type: 'boolean',
              description: 'Receive personal invitations to verify users details'
            },
            email_messages: {
              type: 'boolean',
              description: 'Receive messages from other APPII users'
            },
            email_opportunities: {
              type: 'boolean',
              description: 'Receive emails about potential jobs and opportunities'
            }
          }
        }
      }
    }

    const settings = {
      email_invite_verify: 0,
      email_messages: 1,
      email_opportunities: 1
    }

    it('renders correctly', () => {
      const props = {
        type: 'communications',
        data: settingData,
        settings,
        saveUserSettings: jest.fn(),
        updateUserProfile: jest.fn()
      }

      const tree = renderer.create(
        <SettingsSection {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('when type = privacy', () => {
    const settingData = {
      sections: {
        advertising_preferences: {
          title: 'Advertising preferences',
          settings: {
            cookies_ads: {
              type: 'boolean',
              description: 'APPII may use cookies and similar technologies on third party sites...'
            }
          }
        },
        contact_visibility: {
          title: 'Contact Details',
          settings: {
            visibility_email: {
              type: 'enum',
              enum: { private: 'Only Me', verifiers: 'Verifiers', registered: 'Registered Users', public: 'Everyone' },
              description: 'Who can see your Email Address'
            },
            visibility_twitter: {
              type: 'enum',
              enum: { private: 'Only Me', verifiers: 'Verifiers', registered: 'Registered Users', public: 'Everyone' },
              description: 'Who can see your your Twitter username'
            }
          }
        }
      }
    }

    const settings = {
      visibility_twitter: 'private',
      visibility_email: 'public',
      cookies_ads: 1
    }

    it('renders correctly', () => {
      const user = {
        visibility: 'registered'
      }
      const props = {
        type: 'privacy',
        data: settingData,
        user,
        settings,
        saveUserSettings: jest.fn(),
        updateUserProfile: jest.fn()
      }

      const tree = renderer.create(
        <SettingsSection {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })
})
