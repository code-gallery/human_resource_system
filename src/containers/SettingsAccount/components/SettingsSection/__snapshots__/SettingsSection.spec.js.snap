// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/SettingsAccount/components/SettingsSection when type = communications renders correctly 1`] = `
<div>
  <form>
    <div
      className="form-group"
    >
      <h2
        className="form-group"
      >
        Email Frequency
      </h2>
      <ul>
        <li>
          <div
            className="form-group text-center"
          >
            <span
              className="pull-left settings-item-text"
            >
              Receive personal invitations to verify users details
            </span>
            <div
              className="form-checkbox no-label pull-right"
            >
              <input
                checked={0}
                id="email_invite_verify"
                name="email_invite_verify"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                htmlFor="email_invite_verify"
              >
                 
              </label>
            </div>
          </div>
        </li>
        <li>
          <div
            className="form-group text-center"
          >
            <span
              className="pull-left settings-item-text"
            >
              Receive messages from other APPII users
            </span>
            <div
              className="form-checkbox no-label pull-right"
            >
              <input
                checked={1}
                id="email_messages"
                name="email_messages"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                htmlFor="email_messages"
              >
                 
              </label>
            </div>
          </div>
        </li>
        <li>
          <div
            className="form-group text-center"
          >
            <span
              className="pull-left settings-item-text"
            >
              Receive emails about potential jobs and opportunities
            </span>
            <div
              className="form-checkbox no-label pull-right"
            >
              <input
                checked={1}
                id="email_opportunities"
                name="email_opportunities"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                htmlFor="email_opportunities"
              >
                 
              </label>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </form>
  <button
    className="form-btn save"
    onClick={[Function]}
  >
    Save
  </button>
</div>
`;

exports[`containers/SettingsAccount/components/SettingsSection when type = privacy renders correctly 1`] = `
<div>
  <form>
    <div
      className="form-group"
    >
      <h2
        className="form-group"
      >
        Advertising preferences
      </h2>
      <ul>
        <li>
          <div
            className="form-group text-center"
          >
            <span
              className="pull-left settings-item-text"
            >
              APPII may use cookies and similar technologies on third party sites...
            </span>
            <div
              className="form-checkbox no-label pull-right"
            >
              <input
                checked={1}
                id="cookies_ads"
                name="cookies_ads"
                onChange={[Function]}
                type="checkbox"
              />
              <label
                htmlFor="cookies_ads"
              >
                 
              </label>
            </div>
          </div>
        </li>
      </ul>
    </div>
    <div
      className="form-group"
    >
      <h2
        className="form-group"
      >
        Contact Details
      </h2>
      <ul>
        <li>
          <div
            className="form-group text-center"
          >
            <span
              className="pull-left settings-item-text"
            >
              Who can see your Email Address
            </span>
            <div
              className="pull-right col-xs-12 col-md-3"
            >
              <div
                className="Select has-value is-clearable is-searchable Select--single"
                style={undefined}
              >
                <input
                  disabled={false}
                  name="select-visibility_email"
                  type="hidden"
                  value="public"
                />
                <div
                  className="Select-control"
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  style={undefined}
                >
                  <span
                    className="Select-multi-value-wrapper"
                    id="react-select-2--value"
                  >
                    <div
                      className="Select-value"
                      style={undefined}
                      title={undefined}
                    >
                      <span
                        aria-selected="true"
                        className="Select-value-label"
                        id="react-select-2--value-item"
                        role="option"
                      >
                        Everyone
                      </span>
                    </div>
                    <div
                      className="Select-input"
                      style={
                        Object {
                          "display": "inline-block",
                        }
                      }
                    >
                      <input
                        aria-activedescendant="react-select-2--value"
                        aria-describedby={undefined}
                        aria-expanded="false"
                        aria-haspopup="false"
                        aria-label={undefined}
                        aria-labelledby={undefined}
                        aria-owns=""
                        className={undefined}
                        id={undefined}
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        required={false}
                        role="combobox"
                        style={
                          Object {
                            "boxSizing": "content-box",
                            "width": "5px",
                          }
                        }
                        tabIndex={undefined}
                        value=""
                      />
                      <div
                        style={
                          Object {
                            "height": 0,
                            "left": 0,
                            "overflow": "scroll",
                            "position": "absolute",
                            "top": 0,
                            "visibility": "hidden",
                            "whiteSpace": "pre",
                          }
                        }
                      >
                        
                      </div>
                    </div>
                  </span>
                  <span
                    aria-label="Clear value"
                    className="Select-clear-zone"
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    title="Clear value"
                  >
                    <span
                      className="Select-clear"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "&times;",
                        }
                      }
                    />
                  </span>
                  <span
                    className="Select-arrow-zone"
                    onMouseDown={[Function]}
                  >
                    <span
                      className="Select-arrow"
                      onMouseDown={[Function]}
                    />
                  </span>
                </div>
              </div>
            </div>
          </div>
        </li>
        <li>
          <div
            className="form-group text-center"
          >
            <span
              className="pull-left settings-item-text"
            >
              Who can see your your Twitter username
            </span>
            <div
              className="pull-right col-xs-12 col-md-3"
            >
              <div
                className="Select has-value is-clearable is-searchable Select--single"
                style={undefined}
              >
                <input
                  disabled={false}
                  name="select-visibility_twitter"
                  type="hidden"
                  value="private"
                />
                <div
                  className="Select-control"
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  style={undefined}
                >
                  <span
                    className="Select-multi-value-wrapper"
                    id="react-select-3--value"
                  >
                    <div
                      className="Select-value"
                      style={undefined}
                      title={undefined}
                    >
                      <span
                        aria-selected="true"
                        className="Select-value-label"
                        id="react-select-3--value-item"
                        role="option"
                      >
                        Only Me
                      </span>
                    </div>
                    <div
                      className="Select-input"
                      style={
                        Object {
                          "display": "inline-block",
                        }
                      }
                    >
                      <input
                        aria-activedescendant="react-select-3--value"
                        aria-describedby={undefined}
                        aria-expanded="false"
                        aria-haspopup="false"
                        aria-label={undefined}
                        aria-labelledby={undefined}
                        aria-owns=""
                        className={undefined}
                        id={undefined}
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        required={false}
                        role="combobox"
                        style={
                          Object {
                            "boxSizing": "content-box",
                            "width": "5px",
                          }
                        }
                        tabIndex={undefined}
                        value=""
                      />
                      <div
                        style={
                          Object {
                            "height": 0,
                            "left": 0,
                            "overflow": "scroll",
                            "position": "absolute",
                            "top": 0,
                            "visibility": "hidden",
                            "whiteSpace": "pre",
                          }
                        }
                      >
                        
                      </div>
                    </div>
                  </span>
                  <span
                    aria-label="Clear value"
                    className="Select-clear-zone"
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    title="Clear value"
                  >
                    <span
                      className="Select-clear"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "&times;",
                        }
                      }
                    />
                  </span>
                  <span
                    className="Select-arrow-zone"
                    onMouseDown={[Function]}
                  >
                    <span
                      className="Select-arrow"
                      onMouseDown={[Function]}
                    />
                  </span>
                </div>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
    <div
      className="form-group"
    >
      <h2
        className="form-group"
      >
        Profile visibility
      </h2>
      <ul>
        <div
          className="form-group text-center"
        >
          <span
            className="pull-left settings-item-text"
          >
            Who can see your profile
          </span>
          <div
            className="pull-right col-xs-12 col-md-3"
          >
            <div
              className="Select has-value is-clearable is-searchable Select--single"
              style={undefined}
            >
              <input
                disabled={false}
                name="select-profile-visibility"
                type="hidden"
                value="registered"
              />
              <div
                className="Select-control"
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                style={undefined}
              >
                <span
                  className="Select-multi-value-wrapper"
                  id="react-select-4--value"
                >
                  <div
                    className="Select-value"
                    style={undefined}
                    title={undefined}
                  >
                    <span
                      aria-selected="true"
                      className="Select-value-label"
                      id="react-select-4--value-item"
                      role="option"
                    >
                      Registered Users
                    </span>
                  </div>
                  <div
                    className="Select-input"
                    style={
                      Object {
                        "display": "inline-block",
                      }
                    }
                  >
                    <input
                      aria-activedescendant="react-select-4--value"
                      aria-describedby={undefined}
                      aria-expanded="false"
                      aria-haspopup="false"
                      aria-label={undefined}
                      aria-labelledby={undefined}
                      aria-owns=""
                      className={undefined}
                      id={undefined}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      required={false}
                      role="combobox"
                      style={
                        Object {
                          "boxSizing": "content-box",
                          "width": "5px",
                        }
                      }
                      tabIndex={undefined}
                      value=""
                    />
                    <div
                      style={
                        Object {
                          "height": 0,
                          "left": 0,
                          "overflow": "scroll",
                          "position": "absolute",
                          "top": 0,
                          "visibility": "hidden",
                          "whiteSpace": "pre",
                        }
                      }
                    >
                      
                    </div>
                  </div>
                </span>
                <span
                  aria-label="Clear value"
                  className="Select-clear-zone"
                  onMouseDown={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  title="Clear value"
                >
                  <span
                    className="Select-clear"
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "&times;",
                      }
                    }
                  />
                </span>
                <span
                  className="Select-arrow-zone"
                  onMouseDown={[Function]}
                >
                  <span
                    className="Select-arrow"
                    onMouseDown={[Function]}
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
      </ul>
    </div>
  </form>
  <button
    className="form-btn save"
    onClick={[Function]}
  >
    Save
  </button>
</div>
`;
